### PR TITLE
mirror: Use source time in multi master mode

### PR DESCRIPTION
**Pull request by vadmeste**
Still work in progress.

It works but I need to ensure that it conforms exactly to the specification.

### CODE REVIEW

**Comment by vadmeste** [comment]
So what this change does:
- Forget everything about old sTag & eTag multi master meta attributes
- Each mirror copy will add a source-mod-time-meta-attribute to the target which is equal to the source object modification time
- In addition to the regular diff: comparison between source & target compares mod time in source object and source-mod-time-meta-attribute in target object
- In watch mode, when an object is received and it has the source-mod-time-meta-attribute, do not mirror it.

**Comment by krishnasrinivas** [comment]
> In addition to the regular diff: comparison between source & target compares mod time in source object and source-mod-time-meta-attribute in target object

the comparison algo should be like this:
```
srcActualModTime := latest(srcMeta["Last-Modified"], srcMeta["X-Amz-Meta-Source-Last-Modified"]
dstActualModTime := latest(dstMeta["Last-Modified"], dstMeta["X-Amz-Meta-Source-Last-Modified"]

if srcActualModTime > dstActualModTime {
  doMirro()
}
```

File: cmd/difference.go [hunk 1]
```diff
@@ -59,34 +61,33 @@ func (d differType) String() string {
 	return "unknown"
 }
 
-const multiMasterETagKey = "X-Amz-Meta-Mm-Etag"
-const multiMasterSTagKey = "X-Amz-Meta-Mm-Stag"
+const multiMasterSourceModTime = "X-Amz-Meta-Mm-Mtime"
```
Comment (krishnasrinivas) [inline cmd/difference.go:64]: `X-Amz-Meta-Mm-Source-Mtime` will be better

**Comment by vadmeste** [comment]
@krishnasrinivas could you take a look ?

File: cmd/difference.go [hunk 2]
```diff
@@ -32,13 +32,13 @@ import (
...
-	differInNone     differType = iota // does not differ
-	differInETag                       // differs in ETag
-	differInSize                       // differs in size
-	differInMetadata                   // differs in metadata
-	differInType                       // differs in type, exfile/directory
-	differInFirst                      // only in source (FIRST)
-	differInSecond                     // only in target (SECOND)
+	differInNone          differType = iota // does not differ
+	differInSize                            // differs in size
+	differInMMSourceMTime                   // differes in multi master source mod time
```
Comment (harshavardhana) [inline cmd/difference.go:37]:
```suggestion
	differInMMSourceMTime                   // differs in multi master source mod time
```

File: cmd/difference.go [hunk 3]
```diff
@@ -59,34 +61,66 @@ func (d differType) String() string {
 	return "unknown"
 }
 
-const multiMasterETagKey = "X-Amz-Meta-Mm-Etag"
+const multiMasterSourceModTimeKey = "X-Amz-Meta-Mm-Source-Mtime"
 const multiMasterSTagKey = "X-Amz-Meta-Mm-Stag"
```
Comment (harshavardhana) [inline cmd/difference.go:65]: this should be removed?

File: cmd/difference.go [hunk 4]
```diff
@@ -59,34 +61,66 @@ func (d differType) String() string {
...
+	if t1.Before(t2) {
+		return t1
+	}
+
+	return t2
+}
+
+// multiMasterModTimeUpdated tries to calculate if the object copy in the target
+// is older than the one in the source by comparing the modtime of the data.
+func multiMasterModTimeUpdated(src, dst *ClientContent, multiMasterCtx string) (ret bool) {
```
Comment (harshavardhana) [inline cmd/difference.go:83]: not using the named return why not just keep it `bool`

**Comment by vadmeste** [comment]
@krishnasrinivas could you take a look ?

**Pull request merged**